src/Microsoft.Unity.Analyzers/NullableReferenceTypesSuppressor.cs

```diff
@@ -0,0 +1,166 @@
+﻿/*--------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See LICENSE in the project root for license information.
+ *-------------------------------------------------------------------------------------------*/
+
+//#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.Unity.Analyzers.Resources;
+
+namespace Microsoft.Unity.Analyzers
+{
+	[DiagnosticAnalyzer(LanguageNames.CSharp)]
+	public class NullableReferenceTypesSuppressor : DiagnosticSuppressor
+	{
+		private const string NULLABLE_WARNING = "CS8618";
+
+		internal static SuppressionDescriptor NullableRule { get; } = new SuppressionDescriptor("USP0016", NULLABLE_WARNING, Strings.NullableReferenceTypesSuppressorJustification);
+
+		//this, rather than IA => IA.Create(), should be faster because it doesn't allocate a new IA every time the value is needed
+		public override ImmutableArray<SuppressionDescriptor> SupportedSuppressions { get; } = ImmutableArray.Create(NullableRule);
+
+		public override void ReportSuppressions(SuppressionAnalysisContext context)
+		{
+			foreach (Diagnostic diagnostic in context.ReportedDiagnostics.Where(diagnostic => diagnostic.Id == NULLABLE_WARNING))
+			{
+				SyntaxNode root = diagnostic.Location.SourceTree.GetRoot();
+				SyntaxNode location = root.FindNode(diagnostic.Location.SourceSpan);
+
+				//does the inspected class inherit from Monobehaviour?
+				if (!location.FirstAncestorOrSelf<ClassDeclarationSyntax>()?.BaseList?.ChildNodes().Any(node => GetSymbol(diagnostic.Location.SourceTree, ((BaseTypeSyntax)node).Type, context.Compilation).Extends(typeof(UnityEngine.MonoBehaviour))) ?? true)
```

Comment:
From my understanding, this suppressor can be useful for other types like NetworkBehaviour, StateMachineBehaviour, ScriptableObject... (probably all types using 'messages' as defined [here](https://github.com/microsoft/Microsoft.Unity.Analyzers/blob/main/src/Microsoft.Unity.Analyzers/ScriptInfo.cs#L16)).

So perhaps you can use something like this for the initial type validation:

```csharp
var syntaxTree = diagnostic.Location.SourceTree;
var classDeclaration = location.FirstAncestorOrSelf<ClassDeclarationSyntax>();
if (classDeclaration == null)
	continue;

var typeSymbol = context.GetSemanticModel(syntaxTree).GetDeclaredSymbol(classDeclaration);
var scriptInfo = new ScriptInfo(typeSymbol);
if (!scriptInfo.HasMessages)
	continue;
```

Reply:
I will look into it.

Reply:
I did some research and you were right that this suppressor is applicable for some other types that use messages, not all of them though. I implemented your suggestion and tweaked a few other places in my code.